### PR TITLE
Feature/83 tertiary provider extensions

### PR DESCRIPTION
**Pull request by smudge202**
I'll update this with key info and todo's shortly...

### CODE REVIEW

File: src/Compose/Application.cs [hunk 1]
```diff
@@ -1,19 +1,7 @@
...
-		protected internal virtual void PostServiceConfiguration(IReadOnlyCollection<ServiceDescriptor> services) { }
-    }
-}
\ No newline at end of file
+﻿namespace Compose
+{
+    public class Application : ServiceProvider
+    {
+
+	}
```
Comment (sblackler) [inline src/Compose/Application.cs]: Nit: Aligning of the bracket...

File: src/Compose/ConcurrentTransitionManagerContainer.cs [hunk 2]
```diff
@@ -1,88 +1,88 @@
...
+			= new ConcurrentDictionary<Type, ConcurrentBag<object>>();
+		private static readonly Type Helper = typeof(GenericHelper<>);
+
+		private static class GenericHelper<T>
+		{
+			private static readonly TypeInfo Disposable = typeof(IDisposable).GetTypeInfo();
+			public static void Restore(object untypedManager)
+			{
+				var manager = untypedManager as DynamicRegister<T>;
+				if (Disposable.IsAssignableFrom(manager.CurrentService.GetType().GetTypeInfo()))
```
Comment (sblackler) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: Check for null or use elvis?

File: src/Compose/ConcurrentTransitionManagerContainer.cs [hunk 3]
```diff
@@ -1,88 +1,88 @@
...
+					((IDisposable)manager.CurrentService).Dispose();
```
Comment (sblackler) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: If `CurrentService` always implements IDisposable then we have an unnecessary cast?

File: src/Compose/ConcurrentTransitionManagerContainer.cs [hunk 4]
```diff
@@ -1,88 +1,88 @@
...
+				manager.CurrentService = manager.SnapshotService;
+			}
+
+			public static void Snapshot(object untypedManager)
+			{
+				var manager = untypedManager as DynamicRegister<T>;
+				if (Disposable.IsAssignableFrom(manager.SnapshotService.GetType().GetTypeInfo()))
```
Comment (sblackler) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: As above

File: src/Compose/ConcurrentTransitionManagerContainer.cs [hunk 5]
```diff
@@ -1,88 +1,88 @@
...
+					((IDisposable)manager.SnapshotService).Dispose();
+				manager.SnapshotService = manager.CurrentService;
+			}
+		}
+
+		public void Add<T>(DynamicRegister<T> register)
```
Comment (sblackler) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: Needs an argument check

File: src/Compose/ConcurrentTransitionManagerContainer.cs [hunk 6]
```diff
@@ -1,88 +1,88 @@
...
+				if (managers.TryTake(out manager))
+					restoredManagers.Add(Restore(Helper.MakeGenericType(serviceType).GetTypeInfo(), manager));
+			while (!restoredManagers.IsEmpty)
+				if (restoredManagers.TryTake(out manager))
+					managers.Add(manager);
+		}
+
+		private object Restore(TypeInfo helper, object manager)
+		{
+			helper.DeclaredMethods.Single(x => x.Name == "Restore").Invoke(null, new[] { manager });
```
Comment (sblackler) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: If there is more than one method with the name restore, this will blow up? Maybe worth adding a test around this?

File: src/Compose/ConcurrentTransitionManagerContainer.cs [hunk 7]
```diff
@@ -1,88 +1,88 @@
...
+				if (managers.TryTake(out manager))
+					snapshottedManagers.Add(Snapshot(Helper.MakeGenericType(serviceType).GetTypeInfo(), manager));
+			while (!snapshottedManagers.IsEmpty)
+				if (snapshottedManagers.TryTake(out manager))
+					managers.Add(manager);
+		}
+
+		private object Snapshot(TypeInfo helper, object manager)
+		{
+			helper.DeclaredMethods.Single(x => x.Name == "Snapshot").Invoke(null, new[] { manager });
```
Comment (sblackler) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: Same as line 60.

File: src/Compose/DynamicActivator.cs [hunk 8]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+
+namespace Compose
+{
+	internal sealed class DynamicActivator
+	{
+		private readonly Func<Type> _dynamicTypeFactory;
+		private readonly Func<object> _dynamicManagerFactory;
+
+		public DynamicActivator(Func<Type> dynamicTypeFactory, Func<object> dynamicManagerFactor)
```
Comment (sblackler) [inline src/Compose/DynamicActivator.cs]: nit: naming on last param

File: src/Compose/DynamicActivator.cs [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+		{
+			_dynamicTypeFactory = dynamicTypeFactory;
+			_dynamicManagerFactory = dynamicManagerFactor;
+		}
+
+		public object Create()
+			=> Activator.CreateInstance(_dynamicTypeFactory(), _dynamicManagerFactory());
```
Comment (sblackler) [inline src/Compose/DynamicActivator.cs]: Thought: thinking we can do a wee bit of magic using some expression trees to new up the object instead of using the slowest activation method known to man.

File: src/Compose/DynamicExtensions.cs [hunk 10]
```diff
@@ -1,182 +1,19 @@
...
-}
+﻿using Microsoft.Extensions.DependencyInjection;
+using System;
+using System.Reflection;
+
+namespace Compose
+{
+	public static class DynamicExtensions
+	{
+		internal static Type CreateProxy(this Application app, TypeInfo serviceTypeInfo)
```
Comment (sblackler) [inline src/Compose/DynamicExtensions.cs]: Needs param checking

File: src/Compose/DynamicManagerFactory.cs [hunk 11]
```diff
@@ -1,25 +1,23 @@
...
+using System.Reflection;
+
+namespace Compose
+{
+	internal static class DynamicManagerFactory
+	{
+		private static readonly Type Exposer = typeof(DynamicManagerExposer<,>);
+
+		internal static object ForFactory(TypeInfo dynamicManagerTypeInfo, object dynamicContainer, object transitionManager, object abstractFactory)
+			=> Activator.CreateInstance(Exposer.MakeGenericType(dynamicManagerTypeInfo.GenericTypeArguments),
```
Comment (sblackler) [inline src/Compose/DynamicManagerFactory.cs]: Param checking....

Also, same as https://github.com/smudge202/compose/pull/104/files#diff-a4d0e71969ef85f6a635d25767b267d8R17

File: src/Compose/IlGeneratingDynamicEmitter.cs [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+		private readonly ModuleBuilder _moduleBuilder;
+
+		public IlGeneratingDynamicEmitter()
+		{
+			_assemblyName = CreateAssemblyName();
+			_assemblyBuilder = CreateAssemblyBuilder();
+			_moduleBuilder = CreateModuleBuilder();
+		}
+
+		public Type GetManagedDynamicProxy(TypeInfo serviceTypeInfo)
```
Comment (sblackler) [inline src/Compose/IlGeneratingDynamicEmitter.cs]: Check check check

File: src/Compose/IlGeneratingDynamicEmitter.cs [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+			*/
+			var typeBuilder = _moduleBuilder.DefineType($"{_assemblyName.Name}+{serviceType.FullName}", TypeAttributes.Public | TypeAttributes.Sealed);
+			typeBuilder.AddInterfaceImplementation(serviceType);
+			foreach (var implementedInterface in serviceTypeInfo.ImplementedInterfaces)
+				typeBuilder.AddInterfaceImplementation(implementedInterface);
+			if (serviceTypeInfo.IsGenericType)
+				typeBuilder.AddGenericsFrom(serviceTypeInfo);
+			try
+			{
+				var managerTypeInfo = typeof(DynamicRegister<>).MakeGenericType(serviceType).GetTypeInfo();
```
Comment (sblackler) [inline src/Compose/IlGeneratingDynamicEmitter.cs]: Cache the typeof call?

File: src/Compose/IlGeneratingDynamicEmitter.cs [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+				new CustomAttributeBuilder(typeof (AssemblyCompanyAttribute).GetTypeInfo().DeclaredConstructors.Single(),
+					new object[] {"Devbot.Net"}),
+				new CustomAttributeBuilder(typeof (AssemblyProductAttribute).GetTypeInfo().DeclaredConstructors.Single(),
+					new object[] {"Compose.DynamicProxies"}),
+				new CustomAttributeBuilder(typeof (AssemblyCopyrightAttribute).GetTypeInfo().DeclaredConstructors.Single(),
+					new object[] {$"Devbot.Net {DateTime.Now.Year}"}),
+				new CustomAttributeBuilder(typeof (ComVisibleAttribute).GetTypeInfo().DeclaredConstructors.Single(),
+					new object[] {false}),
+				new CustomAttributeBuilder(typeof (TargetFrameworkAttribute).GetTypeInfo().DeclaredConstructors.Single(),
+					new object[] {".NETFramework,Version=v4.5"})
```
Comment (sblackler) [inline src/Compose/IlGeneratingDynamicEmitter.cs]: Is this framework version still accurate with coreclr etc?

File: src/Compose/ConcurrentTransitionManagerContainer.cs (hunk 2 above)
Comment (smudge202) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: Perhaps. This is a private nested class so less validation of inputs should be required so long as the consumer logic is sound.

However, hopefully this will come out in the wash when Snapshots and Restores are looked at.

File: src/Compose/ConcurrentTransitionManagerContainer.cs (hunk 5 above)
Comment (smudge202) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: I might be wrong, but I think the exposure here as actually internal (not had my first can of the day yet, but iirc the interface is also internal?).

File: src/Compose/ConcurrentTransitionManagerContainer.cs (hunk 6 above)
Comment (smudge202) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: Perhaps. As with above comments, there's probably no point looking at this code in isolation. Need to check with the consumer if this isn't public.

File: src/Compose/DynamicActivator.cs (hunk 9 above)
Comment (smudge202) [inline src/Compose/DynamicActivator.cs]: Bear in mind this "slowest" activation is the one our DI container uses (and all DI containers to my knowledge). :wink: This is part of the dynamic deference, so feel free to take a look but it's a little complicated.

File: src/Compose/DynamicExtensions.cs (hunk 10 above)
Comment (smudge202) [inline src/Compose/DynamicExtensions.cs]: Internal

File: src/Compose/SyncLockDynamicManagerContainer.cs [hunk 15]
```diff
@@ -1,54 +1,54 @@
...
+using System.Reflection;
+
+namespace Compose
+{
+	internal sealed class SyncLockDynamicManagerContainer<Interface, OriginalService> : DynamicManagerContainer<Interface, OriginalService>
+		where Interface : class where OriginalService : Interface
+	{
+		private readonly List<WeakReferencingDynamicManager<Interface, OriginalService>> _managers
+			= new List<WeakReferencingDynamicManager<Interface, OriginalService>>();
+		private readonly object _sync = new object();
```
Comment (sblackler) [inline src/Compose/SyncLockDynamicManagerContainer.cs]: this object is not needed, you can lock directly on `_managers`

File: src/Compose/IlGeneratingDynamicEmitter.cs (hunk 13 above)
Comment (smudge202) [inline src/Compose/IlGeneratingDynamicEmitter.cs]: Actually, we need to have a look at caching here. For a short period (I think due to some weirdness in the xunit runner), this was throwing on a background thread because it was trying to add the same class to the new assembly more than once (complaining the type already exists).

Caching would help with this, but if not, some name collision detection. (Pretty sure the former is more applicable).

Certainly a `private static readonly Type` wouldn't go amiss either as per your comment. :smile:

File: src/Compose/IlGeneratingDynamicEmitter.cs (hunk 14 above)
Comment (smudge202) [inline src/Compose/IlGeneratingDynamicEmitter.cs]: Fair point/question. I'll need to decompile a coreclr assembly and take a peak.

File: src/Compose/TypeInfoExtensions.cs [hunk 16]
```diff
@@ -1,39 +1,32 @@
...
+using System.Linq;
+using System.Reflection;
+
+namespace Compose
+{
+	internal static class TypeInfoExtensions
+	{
+		public static TypeInfo[] GetGenericArguments(this TypeInfo typeInfo)
+		{
+			IEnumerable<Type> arguments = typeInfo.IsGenericTypeDefinition
```
Comment (smudge202) [inline src/Compose/TypeInfoExtensions.cs]: WOnder if there's a reason this isn't `var`.

File: src/Compose/SyncLockDynamicManagerContainer.cs [hunk 17]
```diff
@@ -1,54 +1,54 @@
...
+
+		public void Change(Func<Interface> service)
+		{
+			foreach (var instance in GetActiveManagers())
+				Change(instance, service());
+		}
+
+		private void Change(WeakReferencingDynamicManager<Interface, OriginalService> manager, Interface service)
+		{
+			if (manager.CurrentService != null && Disposable.IsAssignableFrom(manager.CurrentService.GetType().GetTypeInfo()))
```
Comment (sblackler) [inline src/Compose/SyncLockDynamicManagerContainer.cs]: If the new service and the old service are the same implementation type, do we need to dispose? I don't even know if that situation will occur, but just thinking the transition from ServiceA to ServiceA

File: src/Compose/SyncLockDynamicManagerContainer.cs (hunk 15 above)
Comment (smudge202) [inline src/Compose/SyncLockDynamicManagerContainer.cs]: True; old convention I've always used. Happy to change it though.

File: src/Compose/IlGeneratingDynamicEmitter.cs (hunk 13 above)
Comment (sblackler) [inline src/Compose/IlGeneratingDynamicEmitter.cs]: I've seen a number of `typeof()` calls, wondering whether a static `KnowTypes` class would be useful so: `KnownTypes.GenericDynamicRegister` or `KnownTypes.IDisposable`

File: src/Compose/SyncLockDynamicManagerContainer.cs (hunk 17 above)
Comment (smudge202) [inline src/Compose/SyncLockDynamicManagerContainer.cs]: Interesting question. I suspect the answer to that should be dependent on scope. If singleton, it would be bad, but if transient it makes sense.

File: src/Compose/DynamicActivator.cs (hunk 9 above)
Comment (sblackler) [inline src/Compose/DynamicActivator.cs]: Will raise an issue for it, shouldn't be that hard

File: src/Compose/ConcurrentTransitionManagerContainer.cs (hunk 2 above)
Comment (sblackler) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: I see what you're saying, but I think that we should still be checking anyway (everywhere, not just here) as we shouldn't be trusting the logic elsewhere. What if that logic changes?

File: src/Compose/DynamicActivator.cs (hunk 9 above)
Comment (danielcirket) [inline src/Compose/DynamicActivator.cs]: @sblackler would be interested in seeing the difference perf wise seeing as we were recently discussing expression trees.

File: src/Compose/IlGeneratingDynamicEmitter.cs (hunk 13 above)
Comment (smudge202) [inline src/Compose/IlGeneratingDynamicEmitter.cs]: Not a bad idea. I was private scoping them as I saw them, but that works too.

File: src/Compose/DynamicActivator.cs (hunk 9 above)
Comment (smudge202) [inline src/Compose/DynamicActivator.cs]: Expression trees require compiling so I don't imagine they'll be much (if any) quicker. I'm intrigued as to what an expression such as `foo => new Bar(foo)` compiles too (i.e. are you sure it doesn't use activator? :wink: )

File: src/Compose/ConcurrentTransitionManagerContainer.cs (hunk 2 above)
Comment (danielcirket) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: Agree with Stu here, it's easier to just check it and make sure IMO

File: src/Compose/DynamicActivator.cs (hunk 9 above)
Comment (sblackler) [inline src/Compose/DynamicActivator.cs]: Over just one or two instanciations its not worth the complexity, over a prolonged use its worth it: http://ayende.com/blog/3167/creating-objects-perf-implications  As soon as the const of the compile is dealt with, you have comparable performance to `new T()`. We can pay the cost early though so we don't pay it during instantiation :) Couldnt find the original link, but that should show you enough

File: src/Compose/ConcurrentTransitionManagerContainer.cs (hunk 2 above)
Comment (smudge202) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: Not sure if your comments regarding activator performance are contradictory to superfluous argument validation? :smile:

File: src/Compose/DynamicActivator.cs (hunk 9 above)
Comment (smudge202) [inline src/Compose/DynamicActivator.cs]: I would only expect 1 (sometimes a couple) instantiations of each type.

File: src/Compose/ConcurrentTransitionManagerContainer.cs (hunk 2 above)
Comment (smudge202) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: > ""but I think that we should still be checking anyway (everywhere, not just here) as we shouldn't be trusting the logic elsewhere. What if that logic changes?""

Either tests should fail, or we're missing tests?

Comment (danielcirket) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: I still feel that the arguments should be explicitly checked, in many cases I'd imagine that it would make no/little difference being internal and not as open to rubbish arguments being passed in, however there will be cases where it's more useful.

Argument 'theParameter' cannot be null. vs. later of the generic "object not set to an instance of an object", the more defensive and useful error message would be more preferential as a small nitpick - maybe that's just my preference (Yes, I know you can actually just read the rest of the exception info etc etc - but WHATEVER)?

Basically, I can't say you're wrong @smudge202, as tests should cover the scenarios - you're right, but still feel you should check the arguments explicitly :wink: at least you can fail early if required and as above provide a better error message if required than some of the horridly vague framework messages.

I've just re-read this comment and basically it's just waffle and haven't made a concrete argument for or against :disappointed:

Comment (smudge202) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: Interesting. I just discussed this internally and everybody quite quickly came to the decision that the argument validation on internals is unnecessary, and went as far as saying intentionally not argument validating internals should lead to a much less brittle implementation and test coverage of the public API. Probably because I was somewhat bias in my opinion and explained it better in person than I am doing here. :smile: 

I suppose argument exceptions make little difference to the above process, as it can be followed in either case, and as has been mentioned, does produce a nicer exception. However, I'd much rather we eliminate object null references by better management of state and validation of input vs thinking that wrapping those exceptions in an argument null exception will suffice.

Make sense?

Comment (danielcirket) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: Can't disagree with eliminating the cause of the problems in the first place is always a better direction to go in, however I'm not sure how this would make things less brittle and/or impact the test coverage - maybe that's my lack of understanding/knowledge though?

The ANE was just an example, and a poor one at that :cry:

I found I'm rubbish at explaining myself when typing - it's so much easier to discuss verbally as our previous skype calls have found :smile:

Comment (smudge202) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: Actually, I don't think the mention of ANE is incorrect. ONRE's are awful, developers should be fined when they occur (and yes, I make those mistakes too), but I don't think an ANE should be the answer to dealing with ONRE's which I think the OP was suggesting? ANE's of course should be used in the case of validating a public API (where it totally makes sense unless you have some argument validation process cooked into the API like returning a HTTP 300 for example).

By disallowing ANE's on internal methods, it forces the flow be followed up until you reach the public API to resolve an ONRE.

Comment (danielcirket) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: In the ANE case though, assuming that the public API has its validation for the arguments, this doesn't necessarily prevent the ONRE at a lower internal level - it shouldn't happen - we're the ones writing the implementation after all - and realistically should be caught by tests (where applicable), but we all know it happens sometimes.

Your comments have made me consider if maybe I'm being/thinking too defensively even for internal implementations though? 

However I still like the fail early option rather than not - at this point in time at least.

Definitely agree with consequences for ONRE's though, they are horrible! :smile:

Comment (smudge202) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: Yeh, this is actually working through a relatively new philosophy for me also. Previously I'd have written an ANE and thought no more of it. Most developers will frown upon an ONRE much more than they would ARE, in fact, ARE for the most part is probably considered perfectly acceptable. I think ARE's *can* be a perfectly acceptable answer, but not in this case. 

If there had been an ARE here, would anyone have considered the public API level or just accepted it? I suspect the latter; my suggestion isn't an act of coding practice because quite clearly there is plenty of grey area and it probably sways towards ARE. I'm only contemplating the removal of ARE because I think it makes a psychological difference (I actually have a number of practices I follow for this reason such as no squiggles `{ }` on ifs/fors/etc; not necessarily because I think the code output is *better*, but because it invokes a better thought process).

Comment (danielcirket) [inline src/Compose/ConcurrentTransitionManagerContainer.cs]: I would hope that it wouldn't have got to the internal ARE, however if this was a problem, and the ARE was thrown here, I'd be happy - it's internal and hidden from me and it's falling over - internal implementation, so not something I can do - file the bug with examples to reproduce/PR etc.

In addition if the public API was the problem, then the validation/ARE would be shown way before the internals started throwing up problems.

I think in this case though an ARE _is_ acceptable, if the parameter is null then there is a problem and functionally the method(s) can't work.

Very much a large grey area for sure - lot's of opinions and points, and definitely some good constructive discussion :+1:

**Comment by smudge202** [comment]
Pre and Post service configuration added back in. Trickier than before because of the deferred configuration. All tested.

File: src/Compose/DynamicActivator.cs (hunk 9 above)
Comment (sblackler) [inline src/Compose/DynamicActivator.cs]: Going to experiment with this here: https://github.com/sblackler/compose/tree/feature/107-faster-type-activation

**Comment by herecydev** [comment]
Regarding Test Attributes, I've made a [PR](https://github.com/compose-net/compose/pull/109). I'm not sure if you were testing different builds but that package was targeting an earlier beta on dev.

This now works through command line, but the test explorer is all over the place still. Is there any reason why there are beta6 references to [DependencyInjection](https://github.com/aspnet/DependencyInjection) when this targeting RC1U1?

**Comment by smudge202** [comment]
Awesome regarding PR, I'll check it over ASAP. With regards to beta6 references, say whaaaa? Where are you finding beta6 references? I had loads of issues when I migrated this to latest dependencies but found open and closing Visual Studio / Solution solved all.

I did scan the `project.lock.json` after migration to confirm all the packages were rc1u1, did I miss something?

File: src/Compose/IlGeneratingDynamicEmitter.cs (hunk 14 above)
Comment (sblackler) [inline src/Compose/IlGeneratingDynamicEmitter.cs]: Raised as: #110 so that we can track/fix

**Comment by smudge202** [comment]
Guys, do we want to raise issues for all remaining tasks and get this PR into dev? Or would people prefer we resolve a lot of this before it hits dev?

**Comment by herecydev** [comment]
Ah I see now, I forked dev which was still on beta6. This branch is updated (and I probably should have forked off that).

File: src/Compose/SyncLockDynamicManagerContainer.cs (hunk 15 above)
Comment (sblackler) [inline src/Compose/SyncLockDynamicManagerContainer.cs]: Raised as #111

**Comment by smudge202** [comment]
@herecydev Yup, please. :wink: The good news is there's only 20-30 tests using `Fact` that would need changing.

**Comment by sblackler** [comment]
Should I raise a task for bringing over the remaining tests?

**Comment by smudge202** [comment]
@sblackler,  I would be tempted to change the description of the test cleanup issue (#77)

**Comment by herecydev** [comment]
@smudge202 https://github.com/compose-net/compose/pull/112

**Comment by sblackler** [comment]
Have marked TestAttributes as complete thanks to #112

Should probably remove samples from the list considering #113

I'm currently bringing back executable but needing to sort the tests as I go.

**Comment by smudge202** [comment]
Awesome. I'll update OP and do the README(s) then. Would be awesome to shop today/tomorrow...

**Comment by sblackler** [comment]
You happy with the way that i am changing the test names?
